### PR TITLE
Block Library: Register all block with "block.json" on the server'

### PR DESCRIPTION
**Pull request by gziolo**
## Description
This PR registers on the server all core blocks that have `block.json` defined when they don't contain PHP files that would take care of it. This way soon we should be able to get a few benefits like:
1. Ability to expose all registered blocks and their definitions with a new REST API endpoint as proposed in #21065 by @spacedmonkey.
2. Using the server-side preloaded REST API call to the endpoint mentioned in (1) we should be able to finally remove the temporary workaround used here:
  https://github.com/WordPress/wordpress-develop/blob/437ac63d7652b08095dc7f77b6ba4ef52db44dbb/src/wp-admin/edit-form-blocks.php#L111-L115
3. A way to process all core blocks on the server using their definitions.

## How has this been tested?

I checked in the source code that included blocks are listed in the `wp.blocks.unstable__bootstrapServerSideBlockDefinitions` call:

Screen Shot 2020-05-20 at 16 04 47

## Types of changes
 New feature (non-breaking change which adds functionality).

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [N/A] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [N/A] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

**Comment by spacedmonkey** [comment]
@gziolo Do you want to wait until this is merged then I can update my PR at #21065 to expose all the fields in the API?

File: lib/blocks.php
Comment (spacedmonkey) [inline lib/blocks.php:47]: Miss blocks, 
- Code
- Classic
- Column
- Columns
- Cover
- Embed

Comment (gziolo) [reply lib/blocks.php:47]: Good catch, I will print a list of existing blocks and check which I missed and which don't contain `block.json` (at least Embed misses this file).

**Comment by gziolo** [comment]
This is the list of blocks
- [x] Archives (`register_block_type_from_metadata` called in `archives.php` that is required directly)
- [x] Audio
- [x] Block (`block.php`)
- [x] Button
- [x] buttons
- [x] Calendar (`calendar.php`)
- [x] Categories (`categories.php`)
- [ ] Classic
- [ ] Code
- [ ] Column
- [ ] Columns
- [x] Cover (`cover.php`)
- [x] File
- [x] Gallery
- [x] Group
- [x] Heading
- [x] HTML
- [x] Image
- [x] Latest Comments (`latest-comments.php`)
- [x] Latest Posts (`latest-posts.php`)
- [x] Legacy Widget (`legacy-widget.php`)
- [x] List
- [x] Media Text
- [x] Missing
- [x] More
- [x] Navigation (`navigation.php`)
- [x] Navigation Link
- [x] Next Page
- [x] Paragraph
- [x] Post Author (`post-author.php`)
- [x] Post Comments (`post-comments.php`)
- [x] Post Comments Count (`post-comments-count.php`)
- [x] Post Comments Form (`post-comments-form.php`)
- [x] Post Content (`post-content.php`)
- [x] Post Date (`post-date.php`)
- [x] Post Excerpt (`post-excerpt.php`)
- [x] Post Featured Image (`post-featured-image.php`)
- [x] Post Tags (`post-tags.php`)
- [x] Post Title (`post-title.php`)
- [x] Preformatted
- [x] Pullquote
- [x] Query (`query.php`)
- [x] Query Loop (`query-loop.php`)
- [x] Query Pagination (`query-pagination.php`)
- [x] Quote
- [x] RSS (`rss.php`)
- [x] Search (`search.php`)
- [x] Separator
- [x] Shortcode (`shortcode.php`)
- [x] Site Title (`site-title.php`)
- [x] Social Link (`social-link.php`)
- [x] Social Links
- [x] Spacer
- [x] Subhead
- [x] Table
- [x] Tag Cloud (`tag-cloud.php`)
- [x] Template Part (`template-part.php`)
- [x] Text Columns
- [x] Verse
- [x] Video
- [ ] Widget Area?

**Comment by gziolo** [comment]
With 446d52c I added all missing blocks but Embed block. It has to be tackled separately because it's in fact 30-40 blocks that are nearly identical. They should be converted into a single block with multiple variations. @mcsf did you start some work towards that?

**Comment by gziolo** [comment]
I discussed Widgets Area block with @jorgefilipecosta and it seems like it doesn't have to be loaded in every context. I'm sure the same applies to many other blocks added recently. @youknowriad and @mtias do you have some strategy planned moving forward?

@aduth, while working on updates to `block.json` files I realized that some blocks already use `context` implementation, but they don't define `parents` property. Do you this there should be some sort of ancestor relationship enforced to prevent inserting those blocks when they can't read such context?

**Comment by aduth** [comment]
> @aduth, while working on updates to `block.json` files I realized that some blocks already use `context` implementation, but they don't define `parents` property. Do you this there should be some sort of ancestor relationship enforced to prevent inserting those blocks when they can't read such context?

The short answer for me would be "no".

But I do grant that there's an interesting problem here in if and how a block should assume for context to have been provided. Part of the reason that `context` even exists as a property (vs. all available context as some sort of grab bag) is so that the framework can have some awareness of what context a block is expecting, and be able to make accommodations based on that. Specifically I can see this being used in the sort of scenario you describe where framework can deal with the problem that the context needs are not satisfied.

From the block's perspective, it's still an open question if it should assume the context value is available or not, which could be affected by the prior handling. Currently, they do handle this, though in a way that's not especially graceful ([example, returning empty string](https://github.com/WordPress/gutenberg/blob/ce6593477069a64cb13eee15f4d73ebadf53b1e1/packages/block-library/src/post-title/index.php#L18)).

It might be something to observe over time:

- Design-wise, does it ever make sense to implement a graceful fallback rendering in a block where the context is not provided? 
- Otherwise, if we're always doing something naive like returning an empty string, the framework could handle this on behalf of the block instead.
- In the editor, there's a similar question, in how the absence of some context values could be an intentional distinguishing feature (https://github.com/WordPress/gutenberg/pull/22359#issuecomment-628917435), though this too may/is changing as full-site editing develops (#22368)

Finally, it's worth clarifying: Context doesn't necessarily come from the direct parent. It can come anywhere from the ancestry. So `parents` isn't really relevant for `context` in a lot of cases.

**Comment by gziolo** [comment]
**To make it clear for all observers, all that discussion I started with @aduth isn't a blocker to this particular PR.**

@aduth, thank you for your detailed answer. It looks like there are many ways to work with a context in the case when it's not provided in the upper scope of the tree where block lives.

> Finally, it's worth clarifying: Context doesn't necessarily come from the direct parent. It can come anywhere from the ancestry. So `parents` isn't really relevant for `context` in a lot of cases.

I was aware of this limited usage of `parents` property. Still, it raises the question of whether we should be concerned that there is no way to remove from the inserter a block if it has to exist inside another block at any level of nesting. It doesn't necessarily have to be related to the existence of `context` property.

**Comment by mcsf** [comment]
> With [446d52c](https://github.com/WordPress/gutenberg/commit/446d52c26fe591a8b740b8e6ed52526cdd8171ed) I added all missing blocks but Embed block. It has to be tackled separately because it's in fact 30-40 blocks that are nearly identical. They should be converted into a single block with multiple variations. @mcsf did you start some work towards that?

Regrettably no, what little work I had remained in a local branch and I had to deprioritise that task.

**Comment by aduth** [comment]
> One more question, do you envision any issues on the frontend caused by the block registration process happening on the server knowing that we don't run the same logic that hooks operating on `supports` (`align`, `className`, etc.) do on the frontend? (_It's hard to phrase it clearly..._)

I don't really imagine there should be. For the most part, if a block type doesn't have an associated `render_callback`, there's not really much consideration of it (including attributes and supports) in the front-end rendering.

File: lib/blocks.php
Comment (youknowriad) [inline lib/blocks.php:19]: Instead of an explicit array, can we just loop through the folder?

**Review by youknowriad — APPROVED** [review]
Do you think we should create a trac ticket to backport the API and this change?

File: lib/blocks.php
Comment (gziolo) [reply lib/blocks.php:19]: In the ideal world, I would do it, but some blocks are behind the flag and other call `register_block_from_metadata` in PHP file. Well, we could detect the latter, and blacklist the former. Hmmm, I might open a follow-up. I guess I could use `glob` right?

**Comment by gziolo** [comment]
> Do you think we should create a trac ticket to backport the API and this change?

There is still some work going on in https://github.com/WordPress/gutenberg/pull/22491 but I agree that we should at least have a tracking ticker in Trac 👍

**Comment by gziolo** [comment]
Filed a follow-up #22660 for the only remaining block – Embed.

File: lib/blocks.php
Comment (gziolo) [reply lib/blocks.php:19]: There is also this proposed addition to WordPress core from @aduth in https://core.trac.wordpress.org/ticket/49615:
`register_block_type_args` filter

With that, we could find all `block.json` file to register blocks and use this hook to set `render_callback` separately.

**Comment by gziolo** [comment]
Trac ticket creates as well: https://core.trac.wordpress.org/ticket/50263.

**Pull request merged**
